File: apps/desktop/src/database/favorites.ts
import type { Favorite } from "@/types";
import { getStorage } from "@/services/storage";

/** List all favorites, newest first. */
export function listFavorites(): Promise<Favorite[]> {
  return getStorage().listFavorites();
}

/** Add or update a favorite with its baseline value. */
export function addFavorite(itemId: string, baselineValue: number): Promise<void> {
  return getStorage().addFavorite(itemId, baselineValue);
}

/** Remove a favorite by item id. */
export function removeFavorite(itemId: string): Promise<void> {
  return getStorage().removeFavorite(itemId);
}

File: apps/desktop/src-tauri/src/database/favorites.rs
use rusqlite::{params, Connection};

use super::models::Favorite;

/// List favorites, newest first.
pub fn list(conn: &Connection) -> rusqlite::Result<Vec<Favorite>> {
    let mut stmt = conn.prepare(
        "SELECT item_id, baseline_value, created_at
         FROM favorites ORDER BY created_at DESC",
    )?;
    let rows = stmt.query_map([], |row| {
        Ok(Favorite {
            item_id: row.get(0)?,
            baseline_value: row.get(1)?,
            created_at: row.get(2)?,
        })
    })?;
    rows.collect()
}

/// Add (or update the baseline of) a favorite.
pub fn add(conn: &Connection, item_id: &str, baseline_value: f64) -> rusqlite::Result<()> {
    conn.execute(
        "INSERT INTO favorites (item_id, baseline_value)
         VALUES (?1, ?2)
         ON CONFLICT(item_id) DO UPDATE SET baseline_value = excluded.baseline_value",
        params![item_id, baseline_value],
    )?;
    Ok(())
}

/// Remove a favorite by item id.
pub fn remove(conn: &Connection, item_id: &str) -> rusqlite::Result<()> {
    conn.execute("DELETE FROM favorites WHERE item_id = ?1", params![item_id])?;
    Ok(())
}

/// True when the given item is favorited.
pub fn exists(conn: &Connection, item_id: &str) -> rusqlite::Result<bool> {
    let count: i64 = conn.query_row(
        "SELECT COUNT(*) FROM favorites WHERE item_id = ?1",
        params![item_id],
        |row| row.get(0),
    )?;
    Ok(count > 0)
}
